### ISSUE TITLE
surface syntax for arbitrary Exprs, and non-interpolating quote

### ISSUE THREAD

**Issue by JeffBezanson** [JuliaLang/julia#10209]
In some cases we use `Expr`s that the parser is not able to create, leading to ugly printing involving `$(Expr(...))`. The problem is amplified by `quote`, since using interpolating for printing prevents natural printing of quote expressions. For example

```
julia> x = :x
:x

julia> :(:(f($(($x)...))))
:($(Expr(:quote, :(f($(Expr(:$, :((x...,)))))))))
```

And you thought lisp had a lot of parentheses!

My first proposed remedy is to add syntax that lets you specify an Expr head. For example

```
@(foo, x, y)
```

would *parse* as the expression generated by `Expr(:foo, :x, :y)`. This basically lets you write in s-expression syntax, which is a nice escape hatch to have. `@( )` is currently a syntax error.

We also need syntax for non-interpolating quote. A quote that simply returns its argument AST without any interpolation makes it easier to generate quoted expressions, and is especially useful when nesting quotes. In lisp you often see the idiom ``` `',x ```, which means yield the value of x, but quoted, causing exactly the value of `x` to be returned at run time:

```
> (define x '(2 3))
(2 3)

> `',x
'(2 3)
```

With `@( )`, we could write this as `:(@(inert, $x))`. (`inert` is a non-interpolating quote that exists internally but doesn't have syntax yet.)

We can try to think of a more compact syntax for `@(inert,x)`. Maybe `@:x` or `@@x`.

**Comment by johnmyleswhite** [comment]
+1

The gap between surface syntax and attainable AST's has bugged me for a while.

**Comment by StefanKarpinski** [comment]
Should the syntax have commas? It would be more Lispy to write `@(foo x y)`.

**Comment by JeffBezanson** [comment]
How would `@(foo 1 + 1)` parse? Probably best to keep the commas.

**Comment by StefanKarpinski** [comment]
Wouldn't everything inside be S-Expr syntax? So that would be foo applied to 1, the + symbol, and 2.

**Comment by JeffBezanson** [comment]
No, I shouldn't have written "s-expression", because I didn't really mean it. I only want to solve the specific problem of specifying an Expr head. It wouldn't be nice if you had to switch to full s-exprs as soon as you needed that feature. I'd rather have `@(foo, x+y)` than `@(foo (call + x y))`. With full s-exprs, you suddenly need to know the full abstract syntax just to specify one expression head.

**Comment by StefanKarpinski** [comment]
Ah, ok. That makes more sense then. commas it is.
